packages/cache/ARCHITECTURE.md

```diff
@@ -0,0 +1,56 @@
+# TODO: title
+
+## Assumptions
+
+- The `CacheRegistry` will be generated by the "Codegen" system where the full
+  set of query keys are known ahead of time. In Athena's case that will mean
+  that inidivual query ids like `query-id-123` will map to **exactly** the return type of that query.
+  Similarly, `cache.get(someEntityId)` will map to **exactly** the correct type of the "merged model". For an example of generating these type registries, see <https://tsplay.dev/NrnDlN>. This gives strong types for entries whose types are statically known. There is no support for strongly typing dynamic entries -- they will be typed as `unknown` and users will have to typecast or equivalent.
+
+## Features
+
+// TODO: rewrite after writing `save` api
+
+- The cache is serializable. This means e.g. it can be written to and restored from persistent storage. It also means that values must be serializable one way or another.
+
+## Implementation Notes
+
+- 🔥🐉 When cache entries (key,value pairs) are added to the cache (i.e. via `tx.commit()`) particular care is needed.
+  1. The main cache is a `Map<String, WeakRef<ValueType>>`.
```

Comment:
I am a bit confused by `ValueType` here, this is the actual value object mapping to entityUrn that we are storing I assume?

Reply:
Yes correct.

If you look at the TypeScript examples in the Readme, there isn't a `ValueType`, it's usually written as something like
`CacheKeyRegistry[Key]`

e.g. for `get`

```ts
async get<Key extends keyof CacheKeyRegistry>(cacheKey: Key): CacheKeyRegistry[Key] | undefined;
```